Add files to build shared-images-controller image

brianmcarey opened the pull request:
This image is based off of the bootstrap image as the podman go bindings require a running podman system socket to connect to.

Currently this image will just periodically download the latest kubevirtci node images that are required by kubevirt/kubevirt main to the shared images store. It also removes any old images that do not match this latest kubevirtci tag.

The storage.conf is based off of the default config just with the read/write location of container storage updated to match the path of the shared images location.

/cc @xpivarc @dhiller 

Signed-off-by: Brian Carey <bcarey@redhat.com>

brianmcarey commented:
/test check-prow-config

brianmcarey commented:
/rehearse

brianmcarey commented:
/cc @enp0s3

dhiller commented on images/shared-images-controller/main.go:
```diff
@@ -0,0 +1,166 @@
+			log.Infoln("Pulling image: ", name)
+			_, err := images.Pull(ctx, name, nil)
+			if err != nil {
+				log.WithError(err).Errorf("Failed to pull image:", name)
```
Maybe directly return the error, since we are logging it twice, once here and a second time in the main func?

```suggestion
```

xpivarc commented on images/shared-images-controller/main.go:
```diff
@@ -0,0 +1,166 @@
+			log.WithError(err).Errorf("Failure occured when deleting old images")
+		}
+		log.Infof("Waiting for %d hours before checking again", period)
+		time.Sleep(period * time.Hour)
```
When do we kill the job?

xpivarc replied:
Ok, I got confused by the addition of pre/post submit jobs. Do you want to use them for testing? If so would it make sense to implement an option to indicate you want one run and then exit?

brianmcarey replied:
the pre/post submit jobs are just used for building / publishing the image that will be used by the daemonset.

xpivarc replied:
My bad. I see it now. Should have looked before.

brianmcarey commented:
> Mostly nits, please see inline comments.

@dhiller  - thanks for the good feedback. I have addressed your comments here - https://github.com/kubevirt/project-infra/pull/2465/commits/d8b7b72f373e067ee11668b28bce91a9b577bb8b

dhiller commented on images/shared-images-controller/main.go:
```diff
@@ -0,0 +1,156 @@
+	for _, version := range versions {
+		log.Infoln("Kubevirt Provider version: ", version)
+		name := fmt.Sprintf("%sk8s-%s:%s", kubevirtci_repo, version, tag)
+		_, exists := imageNames[name]
+		if exists {
```
nit: you can do this in one line:

```suggestion
		if _, exists := imageNames[name]; exists {
```

dhiller reviewed (approved):
/approve

Looks good to me. One nit inline.

/hold to decide whether you first want to fix the nit

brianmcarey commented:
/rehearse

dhiller commented on images/shared-images-controller/main.go:
```diff
@@ -0,0 +1,155 @@
+	if err != nil {
+		log.WithError(err).Fatalf("Could not connect to podman socket %d", socket)
+	}
+	for {
```
I hope I am not overthinking this, but what if the process gets killed in the middle of an action like deleting images here? Do we need to care about broken image downloads? I guess not, since podman takes care of that? But what about if the deletion is only half way done? Does podman take care of that also?

However, I've been thinking about a more elegant way of running a periodic process, and I've found [Golang Ticker](https://pkg.go.dev/time#Ticker).

I don't think we need to do this now, but maybe in a later step. WDYT?

brianmcarey replied:
> I hope I am not overthinking this, but what if the process gets killed in the middle of an action like deleting images here? Do we need to care about broken image downloads? I guess not, since podman takes care of that? But what about if the deletion is only half way done? Does podman take care of that also?

We would be relying on podman here - podman will try to download the image again when the loop comes around again so it should be ok on the next iteration. We would have to rely on podman for the image deletion too. 
> 
> However, I've been thinking about a more elegant way of running a periodic process, and I've found [Golang Ticker](https://pkg.go.dev/time#Ticker).
> 
> I don't think we need to do this now, but maybe in a later step. WDYT?

That does look good - we could definitely look into that later.

dhiller reviewed (approved):
OK then, please create an issue to use Golang Ticker as discussed in a later iteration.

xpivarc commented:
+1

brianmcarey commented:
/unhold
